libbeat/common/kubernetes/util.go

```diff
@@ -94,54 +93,65 @@ func IsInCluster(kubeconfig string) bool {
 // If host is provided in the config use it directly.
 // If beat is deployed in k8s cluster, use hostname of pod which is pod name to query pod meta for node name.
 // If beat is deployed outside k8s cluster, use machine-id to match against k8s nodes for node name.
-func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (node string) {
+// If node can not be discovered return NODE_NAME env var as default value. In case it is not set return error.
+func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (string, error) {
 	if host != "" {
 		log.Infof("kubernetes: Using node %s provided in the config", host)
-		return host
+		return host, nil
+	}
+	nodeNameEnv := os.Getenv("NODE_NAME")
+	var envError, logerror error
+	if nodeNameEnv == "" {
+		envError = errors.New("kubernetes: NODE_NAME environment variable was not set")
 	}
 	ctx := context.TODO()
 	if inCluster {
 		ns, err := InClusterNamespace()
 		if err != nil {
-			log.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
```

Comment:
Since there is a lot of `if/else` logic, how about covering these cases with some unit tests? There is a [`fake-go-client`](https://github.com/elastic/beats/blob/49e3857fd412ff95d4374118e6e447aff9ebb619/x-pack/elastic-agent/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go#L25) library you can use to simulate any calls to k8s api.